src/proxy/cpsubsystem/AtomicRefProxy.ts

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/** @ignore *//** */
+
+import {HazelcastClient} from '../../HazelcastClient';
+import {BaseCPProxy} from './BaseCPProxy';
+import {IAtomicReference} from '../IAtomicReference';
+import {CPProxyManager} from './CPProxyManager';
+import {RaftGroupId} from './RaftGroupId';
+import {AtomicRefCompareAndSetCodec} from '../../codec/AtomicRefCompareAndSetCodec';
+import {AtomicRefGetCodec} from '../../codec/AtomicRefGetCodec';
+import {AtomicRefSetCodec} from '../../codec/AtomicRefSetCodec';
+import {AtomicRefContainsCodec} from '../../codec/AtomicRefContainsCodec';
+
+/** @internal */
+export class AtomicRefProxy<E> extends BaseCPProxy implements IAtomicReference<E> {
+
+    constructor(client: HazelcastClient,
+                groupId: RaftGroupId,
+                proxyName: string,
+                objectName: string) {
+        super(client, CPProxyManager.ATOMIC_REF_SERVICE, groupId, proxyName, objectName);
+    }
+
+    compareAndSet(expect: E, update: E): Promise<boolean> {
```

Comment:
Does TypeScript supports representing different types with the same type parameter(`E`)? What I am trying to say is that we use `number` and `string` in the code samples(`code_samples/atomic_reference.js`) as parameters. If they must be the same type in the `.ts` code-bases, maybe we should update the code sample

Reply:
In JS code nothing prevents users from doing that. In TS I'd rather mimic the Java interface. If you're a TS user and you want to store `number` and `string` values under the same `AtomicReference`, then it's possible via `getAtomicReference<number | string>`. This call will give you a `AtomicReference<number | string>` which will accept numbers and strings, but nothing else.

Reply:
I misunderstood the question. The answer is yes, TS supports union types.